migration/migration.go

```diff
@@ -19,20 +21,38 @@ import (
 	"github.com/sourcegraph/conc/pool"
 )
 
+type schemaMetadata struct {
+	Version   uint64
+	Done      bool
+	LastState []byte
+}
+
+func (schema *schemaMetadata) fromBytes(data []byte) error {
+	return json.Unmarshal(data, schema)
+}
+
+func (schema schemaMetadata) toBytes() ([]byte, error) {
+	return json.Marshal(schema)
+}
+
 type Migration interface {
-	Before()
-	Migrate(db.Transaction, utils.Network) error
+	Before(key []byte) error
+	Migrate(context.Context, db.Transaction, utils.Network) error
+	State() []byte
```

Comment:
Instead of adding the `State` method, can we make `Migrate` return `[]byte, error`, where the `[]byte` slice is the `IntermediateState`?